letsencrypt/plugins/standalone.py

```diff
@@ -0,0 +1,262 @@
+    """Standalone Authenticator.
+
+    This authenticator creates its own ephemeral TCP listener on the
+    necessary port in order to respond to incoming DVSNI and SimpleHTTP
+    challenges from the certificate authority. Therefore, it does not
+    rely on any existing server program.
+
+    """
+    zope.interface.implements(interfaces.IAuthenticator)
+    zope.interface.classProvides(interfaces.IPluginFactory)
+
+    description = "Standalone Authenticator"
+
+    def __init__(self, *args, **kwargs):
+        super(Authenticator, self).__init__(*args, **kwargs)
+
+        # one self-signed key for all DVSNI and SimpleHTTP certificates
+        self.key = OpenSSL.crypto.PKey()
+        self.key.generate_key(OpenSSL.crypto.TYPE_RSA, bits=2048)
+        # TODO: generate only when the first SimpleHTTP challenge is solved
+        self.simple_http_cert = acme_crypto_util.gen_ss_cert(
+            self.key, domains=["temp server"])
+
+        self.served = collections.defaultdict(set)
+
+        # Stuff below is shared across threads (i.e. servers read
+        # values, main thread writes). Due to the nature of CPython's
+        # GIL, the operations are safe, c.f.
+        # https://docs.python.org/2/faq/library.html#what-kinds-of-global-value-mutation-are-thread-safe
+        self.certs = {}
+        self.simple_http_resources = set()
+
+        self.servers = ServerManager(self.certs, self.simple_http_resources)
+
+    @classmethod
+    def add_parser_arguments(cls, add):
+        add("supported-challenges", help="Supported challenges, "
+            "order preferences are randomly chosen.",
+            type=supported_challenges_validator, default=",".join(
+                sorted(chall.typ for chall in SUPPORTED_CHALLENGES)))
+
+    @property
+    def supported_challenges(self):
+        """Challenges supported by this plugin."""
+        return set(challenges.Challenge.TYPES[name] for name in
+                   self.conf("supported-challenges").split(","))
+
+    def more_info(self):  # pylint: disable=missing-docstring
+        return self.__doc__
+
+    def prepare(self):  # pylint: disable=missing-docstring
+        if any(util.already_listening(port) for port in
+               (self.config.dvsni_port, self.config.simple_http_port)):
+            raise errors.MisconfigurationError(
+                "At least one of the (possibly) required ports is "
+                "already taken.")
+
+    def get_chall_pref(self, domain):
+        # pylint: disable=unused-argument,missing-docstring
+        supported_challenges = self.supported_challenges
+        if not self.config.no_simple_http_tls and not (
+                acme_standalone.ACMETLSServer.SIMPLE_HTTP_SUPPORT):
+            logger.debug("SimpleHTTPS not supported: %s", sys.version)
+            supported_challenges.discard(challenges.SimpleHTTP)
+        chall_pref = list(supported_challenges)
+        random.shuffle(chall_pref)  # 50% for each challenge
```

Comment:
Good point. It might be worthwhile to keep in during beta, but for launch, I think we should have a conversation and determine a predefined ordering to be returned by standalone. If there is no real reason to choose one over the other, I say we just pick an arbitrary ordering and stick to it.